pimd/pim6_cmd.c

```diff
@@ -0,0 +1,452 @@
+/*
+ * PIM for IPv6 FRR
+ * Copyright (C) 2022  Vmware, Inc.
+ *		       Mobashshera Rasool <mrasool@vmware.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+
+#include "lib/json.h"
+#include "command.h"
+#include "if.h"
+#include "prefix.h"
+#include "zclient.h"
+#include "plist.h"
+#include "hash.h"
+#include "nexthop.h"
+#include "vrf.h"
+#include "ferr.h"
+
+#include "pimd.h"
+#include "pim6_cmd.h"
+#include "pim_vty.h"
+#include "lib/northbound_cli.h"
+#include "pim_errors.h"
+#include "pim_nb.h"
+
+#ifndef VTYSH_EXTRACT_PL
+#include "pimd/pim6_cmd_clippy.c"
+#endif
+
+DEFPY (ipv6_pim_joinprune_time,
+       ipv6_pim_joinprune_time_cmd,
+       "ipv6 pim join-prune-interval (1-65535)$jpi",
+       IPV6_STR
+       PIM_STR
+       "Join Prune Send Interval\n"
+       "Seconds\n")
+{
+	char xpath[XPATH_MAXLEN];
+
+	snprintf(xpath, sizeof(xpath), FRR_PIM_ROUTER_XPATH,
+		 "frr-routing:ipv6");
+	strlcat(xpath, "/join-prune-interval", sizeof(xpath));
+
+	nb_cli_enqueue_change(vty, xpath, NB_OP_MODIFY, jpi_str);
+
+	return nb_cli_apply_changes(vty, NULL);
+}
+
+DEFPY (no_ipv6_pim_joinprune_time,
+       no_ipv6_pim_joinprune_time_cmd,
+       "no ipv6 pim join-prune-interval [(1-65535)]",
+       NO_STR
+       IPV6_STR
+       PIM_STR
+       "Join Prune Send Interval\n"
+       IGNORED_IN_NO_STR)
+{
+	char xpath[XPATH_MAXLEN];
+
+	snprintf(xpath, sizeof(xpath), FRR_PIM_ROUTER_XPATH,
+		 "frr-routing:ipv6");
+	strlcat(xpath, "/join-prune-interval", sizeof(xpath));
+
+	nb_cli_enqueue_change(vty, xpath, NB_OP_DESTROY, NULL);
+
+	return nb_cli_apply_changes(vty, NULL);
+}
+
+DEFPY (ipv6_pim_spt_switchover_infinity,
+       ipv6_pim_spt_switchover_infinity_cmd,
+       "ipv6 pim spt-switchover infinity-and-beyond",
+       IPV6_STR
+       PIM_STR
+       "SPT-Switchover\n"
+       "Never switch to SPT Tree\n")
+{
+	const char *vrfname;
```

Comment:
This code looks almost same to similar V4 cmd (ip_pim_spt_switchover_infinity_cmd) except the address-family (frr-routing:ipv6). This comment is common for other PIMv6 CLI cmds as well. Added similar comment in MLD CLI cmds PR as well (https://github.com/FRRouting/frr/pull/10365). I think, we should take a look at PIM/MLD CLI PRs as well.

I think, we should introduce a common file between V4 & V6 and move these code to that file and we can use "#if PIM_IPV == 4" to encode the address-family properly in that common file. That common abstracted function can be called in V4 (pim_cmd.c) & V6 (pim6_cmd.c).

That will help us maintain common code for both V4 & V6 and any common fixes will be go together. This comment is something similar to @donaldsharp 's suggestion on Socket-changes.

Reply:
@rathnasabapathyv : This is only the command installation. If you check further you can see that we have kept the functional code for both ipv4 and ipv6 common, for that you can check the northbound callback APIs. Since the CLIs are different for ipv4 and ipv6 the files are kept separate. Also in future it is easier to add v4 and v6 specific CLIs in separate files. You can see there are multiple v4 CLIs which are currently not added in v6. This seems to be a cleaner approach, hence it is done this way.

Reply:
@mobash-rasool , @patrasar & @SaiGomathiN : 
I understand, we should have separate DEFPY for V4 & V6 cmds, so that CLI cmds for V4 & V6 can be separate and exec can go to write daemon. I am fine to have two files like pim_cmd.c & pim6_cmd.c as well. **My initial comment was not about this.** 

**My suggestion** was to move the content of DEFPY to a common file (for e.g., pim_cmd_cmn.c) into a new wrapper function in that new common file and call that wrapper function in both V4 & V6 DEFPY. That common file can be complied in both V4 & V6 daemon. 

Suggesting something like this. For clarity, using spt-switchover infinity cmd as an e.g.

Since the content of both V4 & V6 DEFPY is exactly same except address-family (frr-routing:ipv4 or frr-routing:ipv6), we can move those contents like this below.

**In pim_nb.h**
```
#if PIM_IPV == 4
#define		FRR_PIM_AF_XPATH_VAL		“frr-routing:ipv4”
#else
#define		FRR_PIM_AF_XPATH_VAL		“frr-routing:ipv6”
#endif
```

**In pim_cmd_cmn.c (this file will be compiled in both V4 & V6 daemon). Suggesting a function name here for reference. Please suggest a better name.**

```
int pim_dephy_wr_pim_spt_switchover_infinity_cmd (struct vty *vty) {
    const char *vrfname;
    char spt_plist_xpath[XPATH_MAXLEN];
    char spt_action_xpath[XPATH_MAXLEN];

    vrfname = pim_cli_get_vrf_name(vty);
    if (vrfname == NULL)
        return CMD_WARNING_CONFIG_FAILED;

    snprintf(spt_plist_xpath, sizeof(spt_plist_xpath),
         FRR_PIM_VRF_XPATH, "frr-pim:pimd", "pim", vrfname,
         FRR_PIM_AF_XPATH_VAL);
    strlcat(spt_plist_xpath, "/spt-switchover/spt-infinity-prefix-list",
        sizeof(spt_plist_xpath));

    snprintf(spt_action_xpath, sizeof(spt_action_xpath),
         FRR_PIM_VRF_XPATH, "frr-pim:pimd", "pim", vrfname,
         FRR_PIM_AF_XPATH_VAL);
    strlcat(spt_action_xpath, "/spt-switchover/spt-action",
        sizeof(spt_action_xpath));

    if (yang_dnode_exists(vty->candidate_config->dnode, spt_plist_xpath))
        nb_cli_enqueue_change(vty, spt_plist_xpath, NB_OP_DESTROY,
                      NULL);
    nb_cli_enqueue_change(vty, spt_action_xpath, NB_OP_MODIFY,
                  "PIM_SPT_INFINITY");

    return nb_cli_apply_changes(vty, NULL);
}
```

**In pim_cmd.c**
```
DEFUN (ip_pim_spt_switchover_infinity,
       ip_pim_spt_switchover_infinity_cmd,
       "ip pim spt-switchover infinity-and-beyond",
       IP_STR
       PIM_STR
       "SPT-Switchover\n"
       "Never switch to SPT Tree\n")
{
    return pim_dephy_wr_pim_spt_switchover_infinity_cmd (vty)
}
```

**In pim6_cmd.c**
```
DEFPY (ipv6_pim_spt_switchover_infinity,
       ipv6_pim_spt_switchover_infinity_cmd,
       "ipv6 pim spt-switchover infinity-and-beyond",
       IPV6_STR
       PIM_STR
       "SPT-Switchover\n"
       "Never switch to SPT Tree\n")
{
    return pim_dephy_wr_pim_spt_switchover_infinity_cmd (vty)
}
```

Reply:
In my opinion, making this common is not required. This way is cleaner. We have already kept the functional code as common.

Reply:
I'm not sure, why do we want to keep a bulk of code duplicate for this case alone, if there is a way for us to reduce the duplication ! We all know the advantage of keeping the code same for V4 & V6 and that's our strategy/design as well for other PRs. If we handle this case additionally, it will be easy for us to maintain the code, instead of repeating the nb-changes/enhancements/fixes for V4 & V6 separately in future for common core functionality attributes. I would like to get a opinion from @donaldsharp & @eqvinox on this. May be, we can discuss this briefly in next sync-up call.

Reply:
I am going to consider this point and will be pushing the changes today.